cmd/snap/cmd_run.go

```diff
@@ -493,30 +483,90 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func isPossibleRaceCondition(oldInfo *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock) bool {
+	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() || hintFlock != nil {
+		// Skip check
+		return false
+	}
+
+	currentInfo, err := snap.ReadCurrentInfo(oldInfo.InstanceName())
+	if err != nil {
+		// Current symlink does not exist, snap is either removed or being refreshed.
+		return true
+	}
+	if currentInfo.SnapRevision() != oldInfo.SnapRevision() {
+		// Snap was refreshed while we did not hold the hint lock.
+		return true
+	}
+
+	return false
+}
+
 func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
 	if x.DebugLog {
 		os.Setenv("SNAPD_DEBUG", "1")
 		logger.Debugf("enabled debug logging of early snap startup")
 	}
 	snapName, appName := snap.SplitSnapApp(snapApp)
-	info, err := getSnapInfo(snapName, snap.R(0))
-	if err != nil {
-		return err
-	}
 
-	app := info.Apps[appName]
-	if app == nil {
-		return fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
-	}
+	var retryCnt int
+	for {
+		if retryCnt > 1 {
+			// This should never happen, but it is better to fail instead
+			// of retrying forever.
+			return fmt.Errorf("internal error: race condition detected, snap-run can only retry once")
+		}
 
-	if !app.IsService() {
-		// TODO: use signal.NotifyContext as context when snap-run flow is finalized
-		if err := maybeWaitWhileInhibited(context.Background(), snapName); err != nil {
+		info, app, hintFlock, err := maybeWaitWhileInhibited(context.Background(), snapName, appName)
+		if err != nil {
 			return err
 		}
-	}
 
-	return x.runSnapConfine(info, app.SecurityTag(), snapApp, "", args)
+		var errRaceConditionDetected = fmt.Errorf("race condition detected")
+		closeFlockOrRetry := func() error {
+			if hintFlock != nil {
+				// It is okay to release the lock here (beforeExec) because snapd unless forced
+				// will not inhibit the snap and do a refresh anymore because it detects app
+				// processes are running for via the established transient cgroup cgroup.
+				hintFlock.Close()
+				return nil
+			}
+			// hintFlock might be nil if the hint file did not exist
+			if isPossibleRaceCondition(info, app, hintFlock) {
+				return errRaceConditionDetected
+			}
+			return nil
+		}
+
+		err = x.runSnapConfine(info, app.SecurityTag(), snapApp, "", closeFlockOrRetry, args)
+		if errors.Is(err, errRaceConditionDetected) {
+			// Let's retry.
+			//
+			// This will not retry infinitely because this can only be caused by
+			// the following cases:
+			// 1.	No hint file existed when we started due to freshly installed snap,
+			//		then a refresh took place and the current symlink was removed in
+			//		the process.
+			//		In the next retry, the hint lock will have been created.
```

Comment:
Could you please remind me where the hint flock is actually created in this case? Iirc this was a point of confusion for me in the original PR, since checking that it was created didn't atomically create it as well.

Reply:
Right, in `cmd/snaplock/runinhibit/inhibit.go`, `flock` is defined as
```go
        flock, err = osutil.OpenExistingLockForReading(HintFile(snapName))
```
So if there is no existing flock, where in the process is it created?

Reply:
The lock is created in `unlink-current-snap` task during refresh

https://github.com/snapcore/snapd/blob/1afc063e510e65f2b66481bbc15dc8a64e4fa9ac/overlord/snapstate/handlers.go#L1452-L1459